lua/diffview/vcs/adapters/hg/init.lua

```diff
@@ -1,18 +1,1141 @@
+      utils.handle_job(j, {
+        fail_on_empty = true,
+        context = context,
+        debug_opt = { no_stdout = true, context = context },
+      })
+
+      if j.code ~= 0 then
+        callback(j:stderr_result() or {}, nil)
+        return
+      end
+
+      local out_status
+
+      if #j:result() == 0 then
+        async.util.scheduler()
+        out_status = vcs_utils.ensure_output(2, { j }, context)
+      end
+
+      if out_status == JobStatus.ERROR then
+        callback(j:stderr_result() or {}, nil)
+        return
+      end
+
+      callback(nil, j:result())
+    end),
+  })
+  -- Problem: Running multiple 'show' jobs simultaneously may cause them to fail
+  -- silently.
+  -- Solution: queue them and run them one after another.
+  vcs_utils.queue_sync_job(job)
+end, 4)
+
+HgAdapter.flags = {
+  ---@type FlagOption[]
+  switches = {
+    { '-f', '--follow', 'Follow renames' },
+    { '-M', '--no-merges', 'List no merge changesets' },
+  },
+  ---@type FlagOption[]
+  options = {
+    { '=r', '--rev=', 'Revspec', prompt_label = "(Revspec)" },
+    { '=l', '--limit=', 'Limit the number of changesets' },
+    { '=u', '--user=', 'Filter on user' },
+    { '=k', '--keyword=', 'Filter by keyword' },
+    { '=b', '--branch=', 'Filter by branch' },
+    { '=B', '--bookmark=', 'Filter by bookmark' },
+    { '=I', '--include=', 'Include files' },
+    { '=E', '--exclude=', 'Exclude files' },
+  },
+}
+
+for _, list in pairs(HgAdapter.flags) do
+  for i, option in ipairs(list) do
+    option = vim.tbl_extend("keep", option, {
+      prompt_fmt = "${label}${flag_name}",
+
+      key = option.key or utils.str_match(option[2], {
+        "^%-%-?([^=]+)=?",
+        "^%+%+?([^=]+)=?",
+      }):gsub("%-", "_"),
+
+      ---@param self FlagOption
+      ---@param value string|string[]
+      render_value = function(self, value)
+        local quoted
+
+        if type(value) == "table" then
+          quoted = table.concat(vim.tbl_map(function(v)
+            return self[2] .. utils.str_quote(v, { only_if_whitespace = true })
+          end, value), " ")
+        else
+          quoted = self[2] .. utils.str_quote(value, { only_if_whitespace = true })
+        end
+
+        return value == "", quoted
+      end,
+
+      ---@param value string|string[]
+      render_default = function(_, value)
+        if value == nil then
+          return ""
+        elseif type(value) == "table" then
+          return table.concat(vim.tbl_map(function(v)
+            v = select(1, v:gsub("\\", "\\\\"))
+            return utils.str_quote(v, { only_if_whitespace = true })
+          end, value), " ")
+        end
+        return utils.str_quote(value, { only_if_whitespace = true })
+      end,
+    })
+
+    list[i] = option
+    list[option.key] = option
+  end
+end
+
+function HgAdapter:is_binary(path, rev)
+  -- TODO
+  return false 
+end
+
+-- TODO: implement completion
+function HgAdapter:rev_completion(arg_lead, opt)
```

Comment:
```suggestion
function HgAdapter:rev_candidates(arg_lead, opt)
```

Reply:
For this function, I'm not sure how to implement.
With the revsets you get some kind of [mini-language](https://hg.mozilla.org/mozilla-central/help/revsets) for filtering, I don't want to write a whole parser for that :sweat_smile:

Reply:
Wow, that's really cool! But, yeah, not need to write a complex parser for that. Completion is only a helpful tool that aims to reduce the friction of typing out a command by hand. Simply listing all branches for instance seems like it would be helpful. And it seems simple to support the most common (?) types of ranges as well, like:

- `<rev>::<rev>`
- `<rev>:<rev>`

The code for this would be really similar to what's currently in the git adapter.

Reply:
Yes, this and the [changeset evolution](https://www.mercurial-scm.org/doc/evolution/) are two things I miss when working with git :) 

I'll add those basic range constructs indeed